### PR TITLE
Add feature to verify SCTs in a certificate with embedded SCTs

### PR DESCRIPTION
**Pull request by primetomas**
This pull requests contains code to enable extracting SCTs embedded in a certificate and verifying those.
It adds a feature to the VerifySignature util to pass null as sct, which then will look in the certificate. It also adds the feature to point "log keys" to a directory containing many log keys in order to verify multiple SCTs embedded in a certificate.

Use:
java -cp target/ctlog-0.1.2-SNAPSHOT-with-deps.jar org.certificatetransparency.ctlog.utils.VerifySignature chain.pem null logkeys

I tested by verifying the github.com TLS cert, using a selections of known log keys. Test files attached.
[logkeys.zip](https://github.com/google/certificate-transparency-java/files/2014532/logkeys.zip)

[chain.txt](https://github.com/google/certificate-transparency-java/files/2014535/chain.txt)

### CODE REVIEW

**Comment by primetomas** [comment]
I fixed it. PrimeKey's CLA was done by tomas@primekey.se, so I added
this email address to my GitHub account.

/Tomas

File: src/main/java/org/certificatetransparency/ctlog/utils/VerifySignature.java [hunk 1]
```diff
@@ -22,33 +34,120 @@ public static void main(String[] args) throws IOException {
...
-
-    Ct.SignedCertificateTimestamp sct;
-    try {
-      sct = Ct.SignedCertificateTimestamp.parseFrom(sctBytes);
-    } catch (InvalidProtocolBufferException e) {
-      System.out.println("Not a protocol buffer. Trying reading as binary");
-      sct = Deserializer.parseSCTFromBinary(new ByteArrayInputStream(sctBytes));
+    if (certs.isEmpty()) {
+      System.out.println("ERROR: Certificates chain does not contain any certificates.");
+      System.exit(-1);
```
[the remark this answers is not in the document]
Comment (primetomas) [reply src/main/java/org/certificatetransparency/ctlog/utils/VerifySignature.java]: Exactly, I just followed the existing practice so it is consistent.

File: src/main/java/org/certificatetransparency/ctlog/LogSignatureVerifier.java [hunk 2]
```diff
@@ -125,41 +132,40 @@ static IssuerInformation issuerInformationFromCertificateIssuer(Certificate cert
...
-      byte[] toVerify = serializeSignedSCTData(leafCert, sct);
-      return verifySCTSignatureOverBytes(sct, toVerify);
+    if (!CertificateInfo.isPreCertificate(leafCert) && !CertificateInfo.hasEmbeddedSCT(leafCert)) {
+	// When verifying final cert without embedded SCTs, we don't need the issuer but can verify directly
+	byte[] toVerify = serializeSignedSCTData(leafCert, sct);
+	return verifySCTSignatureOverBytes(sct, toVerify);
+    }
+    Preconditions.checkArgument(
+        chain.size() >= 2, "Chain with PreCertificate must contain issuer.");
+    // PreCertificate
```
[the remark this answers is not in the document]
Comment (primetomas) [reply src/main/java/org/certificatetransparency/ctlog/LogSignatureVerifier.java]: Updated the java comments. Also changed a parameter name and accompanying javadoc to reflect possibility to use final cert as well. Didn't change the method name though (verifySCTOverPreCertificate)

File: src/main/java/org/certificatetransparency/ctlog/utils/VerifySignature.java [hunk 3]
```diff
@@ -22,33 +34,120 @@ public static void main(String[] args) throws IOException {
...
     }
 
-    System.out.println("Canned SCT: " + sct.toString());
-
-    LogInfo logInfo = LogInfo.fromKeyFile(logPublicKeyFile);
-    LogSignatureVerifier verifier = new LogSignatureVerifier(logInfo);
-    if (verifier.verifySignature(sct, certs)) {
-      System.out.println("Signature verified OK.");
+    // Read log keys
+    Map<String, LogInfo> logInfos = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (primetomas) [reply src/main/java/org/certificatetransparency/ctlog/utils/VerifySignature.java]: Good comment, agree. Done.

File: src/main/java/org/certificatetransparency/ctlog/LogSignatureVerifier.java [hunk 4]
```diff
@@ -125,41 +132,40 @@ static IssuerInformation issuerInformationFromCertificateIssuer(Certificate cert
...
-   * It does not work for verifying a final certificate with the CT extension. TODO(eranm): Add the
-   * ability to remove the CT extension and verify a final certificate.
-   *
    * @param sct SignedCertificateTimestamp received from the log.
    * @param chain The certificates chain as sent to the log.
    * @return true if the log's signature over this SCT can be verified, false otherwise.
    */
   public boolean verifySignature(Ct.SignedCertificateTimestamp sct, List<Certificate> chain) {
-    if (!logInfo.isSameLogId(sct.getId().getKeyId().toByteArray())) {
+    if (sct != null && !logInfo.isSameLogId(sct.getId().getKeyId().toByteArray())) {
```
[the remark this answers is not in the document]
Comment (primetomas) [reply src/main/java/org/certificatetransparency/ctlog/LogSignatureVerifier.java]: Absolutely. Added test, using real world certificate.

**Comment by primetomas** [comment]
Thanks. I think I addressed all (current) review comments.

**Comment by Martin2112** [comment]
I think it's about ready to be merged. Can you bump the version number in pom.xml?

**Comment by primetomas** [comment]
version bumped to 0.1.3-snapshot

**Pull request merged**